### PR TITLE
Ports smart metal foam from FTL

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Adds smart metal foam, which conforms to area borders and walls. It can be made through chemistry by mixing foaming agent, acetone, and iron.
add: Smart metal foam will create foamed plating on tiles exposed to space. Foamed plating can be struck with floor tiles to turn it into regular plating!
/:cl:

Alternative PR to #29946; if reception of this is more favorable, then I'll close it in favor of this one.

10 smart metal foam grenades can be vended from the EngiVend, as well as a full box if you use a coin. It can also be mixed, using 3 parts foaming agent, 1 part acetone, and 1 part iron. 

It functions similarly to normal metal foam, but instead of creating foamed metal on every tile, it will conform to the area it's in. Any space tiles will be made into foamed plating that can be made into normal plating with a floor tile, and if it detects the borders of its area it will block them off with actual meta - thus creating a foamed shape very roughly identical to the basic shape of the room prior to any damage.

### CODE REVIEW

**Comment by DoctoreMad** [comment]
Form the description you gave it, FTL Smart foam is not very different form your previous PR.

**Comment by Xhuis** [comment]
It isn't. Smart foam accomplishes the same thing in a way that's cleaner on code, if less precise. Had I known that it existed, I wouldn't have made the gel to begin with!

**Comment by ike709** [comment]
Porting inflatable walls and doors is also an option.

https://github.com/FTL13/FTL13/pull/780

**Comment by Xhuis** [comment]
I'm not doing that until you also have inflatable bouncy castles, sorry.
(We do have an equivalent of those on /tg/ in the form of holo-firelocks and holofans!)

**Comment by ike709** [comment]
Meh. We have those too, I just _never_ see them used.

**Comment by WJohn** [comment]
Why not just make this a standard mechanic of regular metal foam instead of creating a new version that's just metal foam but a bit better?

Condense things, man!

**Comment by DoctoreMad** [comment]
Well metal foam is kinda shit since its mostly for griefing so I wouldn't disagree with that

**Comment by Xhuis** [comment]
Normal metal foam is more widely-available, and is better for a more imprecise way of fixing holes. I think it still has a place; it's more useful in different ways than smart foam is.

**Comment by Repukan** [comment]
Normal metal foam also doubles as a line of sight blocker, path maker, cover, and hindrance when running.

**Comment by Shadowlight213** [comment]
Should require blended brains as an ingredient

**Comment by Dorsisdwarf** [comment]
Seems off that engineering get 10 grenades of this stuff whereas atmos get 3 grenades of the basically-useless normal foam

**Comment by Xhuis** [comment]
Atmos has atmos resin, I believe.

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 1]
```diff
@@ -68,6 +72,21 @@
 	flick("[icon_state]-disolve", src)
 	QDEL_IN(src, 5)
 
+/obj/effect/particle_effect/foam/smart/kill_foam() //Smart foam adheres to area borders for walls
+	STOP_PROCESSING(SSfastprocess, src)
+	if(metal)
+		var/turf/T = get_turf(src)
+		if(isspaceturf(T)) //Block up any exposed space
+			T.ChangeTurf(/turf/open/floor/plating/foam)
+		T = get_turf(src) //In case we got changed into space, refresh
```
Comment (RemieRichards) [inline code/game/objects/effects/effect_system/effects_foam.dm]: this shouldn't be necessary, turf refs are immutable last I checked.

File: code/game/turfs/simulated/floor/plating.dm [hunk 2]
```diff
@@ -69,3 +70,36 @@
...
+	desc = "Thin, fragile flooring created with metal foam."
+	icon_state = "foam_plating"
+	broken_states = list("foam_plating")
+	burnt_states = list("foam_plating")
+
+/turf/open/floor/plating/foam/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/stack/tile/plasteel))
+		var/obj/item/stack/tile/plasteel/P = I
+		if(P.amount <= 1)
+			to_chat(user, "<span class='danger'>You need at least one floor tile to buil- wait, how did you get a stack with no tiles...?>/span>")
```
Comment (RemieRichards) [inline code/game/turfs/simulated/floor/plating.dm]: No, just wrap the rest under ```if(P.use(1))```

**Comment by Xhuis** [comment]
I have a quick update to push soon, but I'm trying to refine the adaptation of the foam to walls, so it may be a bit.

**Comment by Xhuis** [comment]
* Greatly increased the radius, since in testing the foam grenades were nigh-useless against anything more than a minibomb. It now covers about half of a screen's worth of tiles in foam!
* Did some code stuff to make the code more codey.

File: code/game/turfs/simulated/floor/plating.dm [hunk 3]
```diff
@@ -69,3 +70,34 @@
...
+	desc = "Thin, fragile flooring created with metal foam."
+	icon_state = "foam_plating"
+	broken_states = list("foam_plating")
+	burnt_states = list("foam_plating")
+
+/turf/open/floor/plating/foam/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/stack/tile/plasteel))
+		var/obj/item/stack/tile/plasteel/P = I
+		if(P.use(1))
+			var/obj/L = locate(/obj/structure/lattice, src)
```
Comment (ChangelingRain) [inline code/game/turfs/simulated/floor/plating.dm]: it's locate(/obj/structure/lattice) in src

File: code/game/turfs/simulated/floor/plating.dm [hunk 4]
```diff
@@ -69,3 +70,34 @@
...
+		user.do_attack_animation(src)
+		if(prob(I.force * 20 - 25))
+			user.visible_message("<span class='danger'>[user] smashes through [src]!</span>", \
+							"<span class='danger'>You smash through [src] with [I]!</span>")
+			ChangeTurf(baseturf)
+		else
+			to_chat(user, "<span class='danger'>You hit [src], to no effect!</span>")
+
+/turf/open/floor/plating/foam/ex_act()
+	ChangeTurf(/turf/open/space)
```
Comment (ChangelingRain) [inline code/game/turfs/simulated/floor/plating.dm]: you need a contents explosion in here or it won't explode contents
also, should probably change to baseturf

**Pull request merged**